#include "TaskEditDialog.h"
#include "ui_TaskEditDialog.h"

TaskEditDialog::TaskEditDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::TaskEditDialog)
{
    ui->setupUi(this);
    setupUI();
    connectSignals();
}

TaskEditDialog::TaskEditDialog(const Task& task, QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::TaskEditDialog)
{
    ui->setupUi(this);
    setupUI();
    connectSignals();
    setTaskData(task); // é¢„å¡«å……æ•°æ®
}

TaskEditDialog::~TaskEditDialog()
{
    delete ui;
}

void TaskEditDialog::setupUI()
{
    // è®¾ç½®çª—å£æ ‡é¢˜
    setWindowTitle("ä»»åŠ¡ç¼–è¾‘");

    // è®¾ç½®ä¼˜å…ˆçº§é€‰é¡¹
    ui->priorityComboBox->clear();
    ui->priorityComboBox->addItem("ğŸŸ¢ ä½", Task::Low);
    ui->priorityComboBox->addItem("ğŸŸ¡ ä¸­", Task::Medium);
    ui->priorityComboBox->addItem("ğŸ”´ é«˜", Task::High);

    // è®¾ç½®åˆ†ç±»é€‰é¡¹
    ui->categoryComboBox->clear();
    ui->categoryComboBox->addItem("ğŸ“ æœªåˆ†ç±»");
    ui->categoryComboBox->addItem("ğŸ’¼ å·¥ä½œ");
    ui->categoryComboBox->addItem("ğŸ“š å­¦ä¹ ");
    ui->categoryComboBox->addItem("ğŸ  ç”Ÿæ´»");

    // è®¾ç½®æ—¥æœŸèŒƒå›´
    ui->deadlineDateTimeEdit->setMinimumDateTime(QDateTime::currentDateTime());

    // è®¾ç½®é»˜è®¤æ—¥æœŸä¸ºæ˜å¤©
    ui->deadlineDateTimeEdit->setDateTime(QDateTime::currentDateTime().addDays(1));

    // è®¾ç½®å¯¹è¯æ¡†å¤§å°
    resize(500, 400);
}

void TaskEditDialog::connectSignals()
{
    connect(ui->saveButton, &QPushButton::clicked, this, &TaskEditDialog::onSaveClicked);
    connect(ui->cancelButton, &QPushButton::clicked, this, &TaskEditDialog::onCancelClicked);
}

void TaskEditDialog::setTaskData(const Task& task)
{
    ui->titleLineEdit->setText(task.getTitle());
    ui->descriptionTextEdit->setPlainText(task.getDescription());
    ui->priorityComboBox->setCurrentIndex(task.getPriority());
    ui->deadlineDateTimeEdit->setDateTime(task.getDeadline());
    ui->categoryComboBox->setCurrentText(task.getCategory());
}

Task TaskEditDialog::getTaskData() const
{
    Task task;
    task.setTitle(ui->titleLineEdit->text());
    task.setDescription(ui->descriptionTextEdit->toPlainText());
    task.setPriority(static_cast<Task::Priority>(
        ui->priorityComboBox->currentData().toInt()));
    task.setDeadline(ui->deadlineDateTimeEdit->dateTime());
    task.setCategory(ui->categoryComboBox->currentText());
    return task;
}

void TaskEditDialog::onSaveClicked()
{
    // éªŒè¯å¿…å¡«å­—æ®µ
    if (ui->titleLineEdit->text().trimmed().isEmpty()) {
        QMessageBox::warning(this, "è¾“å…¥é”™è¯¯", "ä»»åŠ¡æ ‡é¢˜ä¸èƒ½ä¸ºç©ºï¼");
        return;
    }

    accept(); // å…³é—­å¯¹è¯æ¡†å¹¶è¿”å› Accepted
}

void TaskEditDialog::onCancelClicked()
{
    reject(); // å…³é—­å¯¹è¯æ¡†å¹¶è¿”å› Rejected
}
